cmd/gateway-b2-anonymous.go

```diff
@@ -0,0 +1,144 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// mkRange converts offset, size into Range header equivalent.
+func mkRange(offset, size int64) string {
+	if offset == 0 && size == 0 {
+		return ""
+	}
+	if size == 0 {
+		return fmt.Sprintf("%s%d-", byteRangePrefix, offset)
+	}
+	return fmt.Sprintf("%s%d-%d", byteRangePrefix, offset, offset+size-1)
+}
+
+// AnonGetObject - Get object anonymously
+func (l *b2Objects) AnonGetObject(bucket string, object string, startOffset int64, length int64, writer io.Writer) error {
+	uri := fmt.Sprintf("%s/file/%s/%s", l.b2Client.DownloadURI, bucket, object)
+	req, err := http.NewRequest("GET", uri, nil)
+	if err != nil {
+		return err
+	}
+	rng := mkRange(startOffset, length)
+	if rng != "" {
+		req.Header.Set("Range", rng)
+	}
+	resp, err := l.anonClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(writer, resp.Body)
+	return err
+}
+
+// Converts http Header into ObjectInfo. This function looks for all the
+// standard Backblaze B2 headers to convert into ObjectInfo.
+//
+// Content-Length is converted to Size.
+// X-Bz-Upload-Timestamp is converted to ModTime.
+// X-Bz-Info-<header>:<value> is converted to <header>:<value>
+// Content-Type is converted to ContentType.
+// X-Bz-Content-Sha1 is converted to ETag.
+func headerToObjectInfo(bucket, object string, header http.Header) (objInfo ObjectInfo, err error) {
+	clen, err := strconv.ParseInt(header.Get("Content-Length"), 10, 64)
+	if err != nil {
+		return objInfo, err
+	}
+	timeStamp, err := strconv.ParseInt(header.Get("X-Bz-Upload-Timestamp"), 10, 64)
+	if err != nil {
+		return objInfo, err
+	}
+
+	info := make(map[string]string)
+	for key := range header {
+		if !strings.HasPrefix(key, "X-Bz-Info-") {
+			continue
+		}
+		name, err := url.QueryUnescape(strings.TrimPrefix(key, "X-Bz-Info-"))
+		if err != nil {
+			return objInfo, err
+		}
+		val, err := url.QueryUnescape(header.Get(key))
+		if err != nil {
+			return objInfo, err
+		}
+		info[name] = val
+	}
+	objInfo = ObjectInfo{
+		Bucket:      bucket,
+		Name:        object,
+		ContentType: header.Get("Content-Type"),
+		ModTime:     time.Unix(timeStamp/1000, timeStamp%1000*1e6),
```

Comment:
Yes but why is your approach more idiomatic? @krisis .

Reply:
It is idiomatic since it uses standard Go types like time.Duration to work with milliseconds. To work with milliseconds available as string in header, we could do the following,
```go
timeStamp, err := time.ParseDuration(header.Get("X-Bz-Upload-Timestamp"))
if err != nil {
}
modTime := time.Unix(0, 0).Add(timeStamp)
```
This way we use the appropriate Go types to work with time in milliseconds instead of dropping down to low-level types like int64, int.